### PR TITLE
Upgrade the admin server during Cloud 6-7 Upgrade

### PR DESCRIPTION
**Pull request by jsuchome**
mkcloud will need to use new upgrade methods for Cloud6-7 upgrade. 

Requires 
- ~~https://github.com/crowbar/crowbar-init/pull/32~~
- ~~https://github.com/crowbar/crowbar-core/pull/618~~
- ~~https://github.com/crowbar/crowbar-openstack/pull/514~~

### CODE REVIEW

File: scripts/mkcloud [hunk 1]
```diff
@@ -1089,9 +1089,20 @@ function onhost_reset_admin()
...
-    crowbarrestore "with_upgrade"
+    if iscloudver 6plus ; then
+      # FIXME we need to prepare Cloud7 + SP2 repos now: use parts of onadmin_prepare_cloudupgrade?
+      onadmin upgrade_admin_server
+      wait_for 50 3 "! nc -z $adminip 22" 'crowbar to go down after upgrade'
+      wait_for_crowbar_ssh
+      # FIXME wait until it is back online with admin-server-upgraded-ok
+
+      # use crowbar-init to bootstrap crowbar
+      want_postgresql=1 bootstrapcrowbar 1
```
Comment (MaximilianMeister) [inline scripts/mkcloud]: I think it should be
```
want_postgresql=1 onadmin bootstrapcrowbar 1
```

**Comment by jsuchome** [comment]
The execution of upgrade action would be better via crowbarctl, once it has support for that (in 3.0!)

**Comment by MaximilianMeister** [comment]
> The execution of upgrade action would be better via crowbarctl, once it has support for that (in 3.0!)

right, we have to wait until https://github.com/crowbar/crowbar-client/pull/96 is merged. I have tested it already in combination with https://github.com/crowbar/crowbar-core/pull/604 and it seems to work fine

**Comment by MaximilianMeister** [comment]
```
➜  crowbar-client git:(upgrade) ✗ crowbarctl upgrade crowbar
Triggered Crowbar operating system upgrade
```

**Comment by jsuchome** [comment]
Switched to `crowbartcl` usage

**Comment by jsuchome** [comment]
So, this finally works for me. `mkcloud cloudupgrade` moves admin server to Cloud7 running on SP2 (all repositories set), nodes are in crowbar-upgrade state.
I would leave the next step (upgrading the nodes) unimplemented as it relies on a work that is not done yet.

File: scripts/mkcloud [hunk 2]
```diff
@@ -1091,14 +1091,28 @@ function onhost_reset_admin()
 function crowbarupgrade_5plus()
 {
     onadmin prepare_crowbar_upgrade
-    crowbarbackup "with_upgrade"
-    export cloudsource=$upgrade_cloudsource
-    crowbarrestore "with_upgrade"
+    if iscloudver 6plus ; then
+      onadmin prepare_cloudupgrade_repos_6_to_7
```
Comment (MaximilianMeister) [inline scripts/mkcloud]: indentation needs to be fixed

**Comment by MaximilianMeister** [comment]
> So, this finally works for me. mkcloud cloudupgrade moves admin server to Cloud7 running on SP2 (all repositories set), nodes are in crowbar-upgrade state.

where do you actually set the state of the nodes to upgrading? you can do that now with: `crowbarctl upgrade prepare`

**Comment by jsuchome** [comment]
> where do you actually set the state of the nodes to upgrading? you can do that now with: 
> crowbarctl upgrade prepare

This is `onadmin prepare_crowbar_upgrade` call, still used the same way as for Cloud5

**Comment by rsalevsky** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;uUsgpcsp&#x2F;62-5-s46p11-move-from-reinstall-to-os-upgrade-for-admin-server"> (5) S46P11: Move from reinstall to OS upgrade for admin server</a>

File: scripts/qa_crowbarsetup.sh [hunk 3]
```diff
@@ -1740,6 +1740,21 @@ function install_crowbar_init()
...
 
+function onadmin_finalize_server_upgrade()
+{
+  if iscloudver 5minus; then
+      complain 11 "This upgrade path is only supported for Cloud 6+"
+  fi
+
+  # make sure that crowbar schemas are up to date
+  pushd /opt/crowbar/crowbar_framework
+  su -s /bin/sh - crowbar sh -c "RAILS_ENV=development rake crowbar:schema_migrate_prod"
```
Comment (rsalevsky) [inline scripts/qa_crowbarsetup.sh]: Did you try already to run db:migrate after the zypper dup upgrade?

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: Not yet, just about to do it...

**Comment by jsuchome** [comment]
> I would leave the next step (upgrading the nodes) unimplemented as it relies on a work that is not 
> done yet.

UPDATE: the node upgrades seems to work after latest changes

File: scripts/qa_crowbarsetup.sh [hunk 4]
```diff
@@ -1740,8 +1740,24 @@ function install_crowbar_init()
...
 
+function onadmin_finalize_server_upgrade()
+{
+    if iscloudver 5minus; then
+        complain 11 "This upgrade path is only supported for Cloud 6+"
+    fi
+
+    # make sure that crowbar schemas are up to date
+    pushd /opt/dell/crowbar_framework
+    sudo -u crowbar RAILS_ENV=production bin/rake crowbar:schema_migrate_prod
```
Comment (rsalevsky) [inline scripts/qa_crowbarsetup.sh]: We should move this to crowbar-init as new endpoint.

File: scripts/qa_crowbarsetup.sh [hunk 5]
```diff
@@ -1754,6 +1770,15 @@ function onadmin_bootstrapcrowbar()
...
+        if [[ $migrate_db = 1 ]] ; then
+            http_code=`curl -s -X POST -H "Accept: application/vnd.crowbar.v2.0+json" -o migrate-database.txt -w '%{http_code}' $crowbar_init_api/database/migrate`
+            if ! [[ $http_code =~ [23].. ]] ; then
+                cat migrate-database.txt
+                complain 36 "Could not migrate the Crowbar database"
+            fi
+        fi
+
         # FIXME: when https://github.com/SUSE-Cloud/automation/pull/1105 is merged
         http_code=`curl -s -X POST -H "Accept: application/vnd.crowbar.v2.0+json" -o initialize-crowbar.txt -w '%{http_code}' $crowbar_init_api/init`
```
Comment (rsalevsky) [inline scripts/qa_crowbarsetup.sh]: Not sure, but I think we don't need to run install-chef-suse after the upgrade. It's already installed...

File: scripts/qa_crowbarsetup.sh (hunk 4 above)
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: It's true it won't be needed only by mkcloud. Can't you make it mart of db migrate call?

File: scripts/qa_crowbarsetup.sh (hunk 5 above)
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: OK, just tell me what needs to be done here

**Comment by jsuchome** [comment]
@rsalevsky I've added new migrate call for the schemas

**Comment by MaximilianMeister** [comment]
>> where do you actually set the state of the nodes to upgrading? you can do that now with: 
>> crowbarctl upgrade prepare

> This is onadmin prepare_crowbar_upgrade call, still used the same way as for Cloud5

can you update it there as well? https://github.com/SUSE-Cloud/automation/blob/master/scripts/qa_crowbarsetup.sh#L4468 doesn't apply anymore, and then we have also some testing for `crowbarctl`

**Comment by jsuchome** [comment]
I've restarted the gating job with
```
mkcloudtarget=plain_with_upgrade
cloudsource=develcloud6
upgradecloudsource=develcloud7
```

**Comment by MaximilianMeister** [comment]
@jsuchome hmm.. mkcloud failed too early, it hasn't run into the upgrade codepath it seems.

**Comment by jsuchome** [comment]
Merging. If they are problems while installing the nodes, let's solve them later.

**Pull request merged**